### PR TITLE
[DSCP-59] Define private blockchain types using Snowdrop

### PR DESCRIPTION
**Pull request by flyingleafe**
- [x] Add Snowdrop as a dependency and migrate to LTS-10.8
- [x] Define private block datatype
- [x] Define private transaction datatype
  - [x] Define datatypes and interfaces for hashing and signing

### CODE REVIEW

**Comment by Heimdell** [comment]
Wow. I doesn't see Adga in my package indices on Ubuntu. I'll set it up manually, but we need to find what dependency references it.

**Comment by flyingleafe** [comment]
@Heimdell ha-ha, welcome.
For me, installing Agda system-wide solved the problem.

**Comment by Heimdell** [comment]
For me, it didn't, since terminals in linux honor `--pure` option. I had to kill `~/.stack/indices/Hackage` and now it is installing.

**Comment by Heimdell** [comment]
It builds on archlinux (after injecting jemalloc-4.5 instead of 5.0 into `nix`).

**Comment by Heimdell** [comment]
In builds on ubuntu with the same jemalloc injection.

**Review by Heimdell — APPROVED** [review]
It builds, tests are succeeding.

**Comment by yorickvP** [comment]
We're working on supporting different ghc versions and supporting ssh dependencies. `cardano-sl-networking` fails to build with the following log: https://gist.github.com/yorickvP/da06129f5ec3083fe9527af76e342d18

**Comment by flyingleafe** [comment]
I managed to pin exact `jemalloc` version (4.5.0) in `shell.nix`. Unfortunately, it means that at first build `jemalloc-4.5.0` itself should be built, which takes plenty of time.

**Comment by flyingleafe** [comment]
During thinking about transaction types, I understood that these types should necessarily include hashes and signatures, and couldn't help but implement hashes and signatures in a way which can be later abstracted to Lootbox.

**Comment by flyingleafe** [comment]
Seems enough. @Heimdell @lambda76 please take a look.

File: shell.nix [hunk 1]
```diff
@@ -0,0 +1,32 @@
+with import <nixpkgs> {
+  config = {
+    packageOverrides = pkgs: {
+      jemalloc = pkgs.jemalloc.overrideAttrs (oldAttrs: rec {
```
Comment (lambda76) [inline shell.nix]: Might be better to pin exact version of nixpkgs instead of pinning per package.

**Review by lambda76 — APPROVED** [review]

File: shell.nix (hunk 1 above)
Comment (flyingleafe) [reply shell.nix]: Unfortunately, in the older `nixpkgs` release which contained `jemalloc-4.5.0` there was no `ghc-8.2.2`, used in `lts-10.8`, which caused numerous other problems.

**Pull request merged**